packages/plugin-core/src/commands/MoveHeader.ts

```diff
@@ -0,0 +1,311 @@
+import {
+  DendronError,
+  NoteProps,
+  NoteUtils,
+  VaultUtils,
+  getSlugger,
+} from "@dendronhq/common-all";
+import {
+  HistoryService,
+  DendronASTDest,
+  DendronASTTypes,
+  AnchorUtils,
+  Processor,
+  RemarkUtils,
+  MDUtilsV5,
+  ProcMode,
+  visit,
+  Heading,
+  Node,
+} from "@dendronhq/engine-server";
+import { LookupControllerV3 } from "../components/lookup/LookupControllerV3";
+import {
+  NoteLookupProvider,
+  NoteLookupProviderSuccessResp,
+} from "../components/lookup/LookupProviderV3";
+import { DENDRON_COMMANDS } from "../constants";
+import { VSCodeUtils } from "../utils";
+import { BasicCommand } from "./base";
+import _ from "lodash";
+import path from "path";
+import { getEngine } from "../workspace";
+import { EngineAPIService } from "../services/EngineAPIService";
+import { delayedUpdateDecorations } from "../features/windowDecorations";
+import { findReferences } from "../utils/md";
+
+type CommandInput =
+  | {
+      dest?: NoteProps;
+    }
+  | undefined;
+type CommandOpts = {
+  origin: NoteProps;
+  nodesToMove: Node[];
+  modifiedOriginTree: Node;
+  originProc: Processor;
+  engine: EngineAPIService;
+} & CommandInput;
+type CommandOutput = {
+  updated: NoteProps[];
+} & CommandOpts;
+
+export class MoveHeaderCommand extends BasicCommand<
+  CommandOpts,
+  CommandOutput
+> {
+  key = DENDRON_COMMANDS.MOVE_HEADER.key;
+
+  headerNotSelectedError = new DendronError({
+    message: "You must first select the header you want to move.",
+  });
+
+  noActiveNoteError = new DendronError({
+    message: "No note open.",
+  });
+
+  getProc = (engine: EngineAPIService, note: NoteProps) => {
+    return MDUtilsV5.procRemarkParse(
+      { mode: ProcMode.FULL },
+      {
+        engine,
+        fname: note.fname,
+        vault: note.vault,
+        dest: DendronASTDest.MD_DENDRON,
+      }
+    );
+  };
+
+  /**
+   * Recursively check if two given node is identical.
+   * At each level _position_ is omitted as this can change if
+   * you are comparing from two different trees.
+   * @param a first {@link Node} to compare
+   * @param b second {@link Node} to compare
+   * @returns boolean
+   */
+  hasIdenticalChildren = (a: Node, b: Node): boolean => {
```

Comment:
is this expensive for big documents? will it be worthwile to add a shallow flag where we only check 2 levels deep?

Reply:
This only runs when we hit a heading while traversing, which has a pretty bounded depth.

(note that any content that follows a heading is not a children of the heading. they are siblings.)

```md
# **~~deep header~~** foo **_bar_** baz [[wiki]] 

Some paragraph
```

Here the part of the heading that has nested formatting (bold, and then strikethrough) will have the worst case depth of 4. (heading > emphasis > delete > text)

A heading can have a Phrasing Content as children, I don't see anything that can have an unbounded depth here.

That, and the fact that the iteration stops and short circuits immediately once it hits a falsy item make me think it won't be too expensive.

The absolute worst case scenario I can think of is when a person has a gigantic file, with a bunch of headings where every single heading has 4 levels of depth of markdown formatting, and targets the very last heading to move.

Even then, the depth is bounded to be pretty shallow. So I don't think we need to worry about this being expensive.